### ISSUE TITLE
Provide cpp module

### ISSUE THREAD

**Issue by henryiii** [scikit-hep/boost-histogram#155]
There are some pythonisations (mostly related to arrays and indexing) that duplicate the C++ API. We could add a `boost_histogram.cpp` module, and this would provide the exact Boost.Histogram API where possible.

Things to move or provide there:

```python
.at(i) # In Python: [] and bh.underflow/overflow
.axis(i) # Tuple `.axes` in Python
```

This might be supportable alongside `boost_histogram`, or it might inject the methods directly to the existing classes. I'm not sure I'm very excited about creating a new collection of classes -- but if it can be done carefully, it might work (Basically we need Ruby's `extend`...).

**Comment by henryiii** [comment]
@HDembinski, would you be okay with me hiding `.at -> `._at` and `.axis` -> `._axis` in 0.5.1 so that no one starts depending on these in the main package?

**Comment by henryiii** [comment]
Corrected plan: In 0.5.1, the `.at` and `.axis` would still exist, but would raise a DepricationWarning and then call `._at` and `._axis`. In 0.6.0, these would be added or made available in compat, without a warning, and would disappear from the non-compat version.

**Comment by HDembinski** [comment]
That is fine!
